MmSupervisorPkg/Test/MmSupvRequestUnitTestApp/MmSupvMemPolicyUnitTest.c

```diff
@@ -0,0 +1,479 @@
+        (*BaseAddress)[VtdIndex] = ((EFI_ACPI_DMAR_DRHD_HEADER *)DmarHeader)->RegisterBaseAddress;
+        (*Size)[VtdIndex]        = EFI_PAGE_SIZE;
+        VtdIndex++;
+        break;
+
+      default:
+        break;
+    }
+
+    DmarHeader = (EFI_ACPI_DMAR_STRUCTURE_HEADER *)((UINTN)DmarHeader + DmarHeader->Length);
+  }
+
+  Status = EFI_SUCCESS;
+
+Done:
+  return Status;
+}
+
+/*
+  Helper function to get all IOMMU controller regions from IVRS table
+*/
+EFI_STATUS
+GetIommuBaseAmd (
+  OUT UINT64  **BaseAddress,
+  OUT UINT64  **Size,
+  OUT UINTN   *Count
+  )
+{
+  EFI_ACPI_IVRS_HEADER  *AcpiIVRSTable;
+  IVHD_Header           *IvhdHeader;
+  EFI_STATUS            Status;
+  UINTN                 IommuIndex;
+
+  // First get IVRS table
+  AcpiIVRSTable = NULL;
+  Status        = GetAcpiTable (EFI_ACPI_6_3_IO_VIRTUALIZATION_REPORTING_STRUCTURE_SIGNATURE, (VOID **)&AcpiIVRSTable);
+  if (EFI_ERROR (Status)) {
+    goto Done;
+  }
+
+  // Then get through the table to get a count of IVHDs
+  IommuIndex = 0;
+  IvhdHeader = (IVHD_Header *)((UINTN)(AcpiIVRSTable + 1));
+  while ((UINTN)IvhdHeader < (UINTN)AcpiIVRSTable + AcpiIVRSTable->Header.Length) {
+    switch (IvhdHeader->Type) {
+      case IVHD_TYPE_10H:
+      case IVHD_TYPE_11H:
+      case IVHD_TYPE_40H:
+        IommuIndex++;
+        break;
+      default:
+        break;
+    }
+
+    IvhdHeader = (IVHD_Header *)((UINTN)IvhdHeader + IvhdHeader->Length);
+  }
+
+  *BaseAddress = AllocatePool (IommuIndex * sizeof (UINT64));
+  if (BaseAddress == NULL) {
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  *Size = AllocatePool (IommuIndex * sizeof (UINT64));
+  if (Size == NULL) {
+    FreePool (*BaseAddress);
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  *Count = IommuIndex;
+
+  IommuIndex = 0;
+  IvhdHeader = (IVHD_Header *)((UINTN)(AcpiIVRSTable + 1));
+  while ((UINTN)IvhdHeader < (UINTN)AcpiIVRSTable + AcpiIVRSTable->Header.Length) {
+    switch (IvhdHeader->Type) {
+      case IVHD_TYPE_10H:
+      case IVHD_TYPE_11H:
+      case IVHD_TYPE_40H:
+        (*BaseAddress)[IommuIndex] = IvhdHeader->IOMMUBaseAddress;
+        (*Size)[IommuIndex]        = EFI_PAGE_SIZE;
+        IommuIndex++;
+        break;
+
+      default:
+        break;
+    }
+
+    IvhdHeader = (IVHD_Header *)((UINTN)IvhdHeader + IvhdHeader->Length);
+  }
+
+  Status = EFI_SUCCESS;
+
+Done:
+  return Status;
+}
+
+/*
+  Helper function to get all TXT related regions for Intel processors
+*/
+EFI_STATUS
+GetTxtRegions (
+  OUT UINT64  **BaseAddress,
+  OUT UINT64  **Size,
+  OUT UINTN   *Count
+  )
+{
+  UINTN   TxtIndex;
+  UINT32  Temp;
+
+  // TXT public and private regions + Heap + DPR
+  *BaseAddress = AllocatePool (TXT_REGION_COUNT * sizeof (UINT64));
+  if (BaseAddress == NULL) {
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  *Size = AllocatePool (TXT_REGION_COUNT * sizeof (UINT64));
+  if (Size == NULL) {
+    FreePool (*BaseAddress);
+    return EFI_OUT_OF_RESOURCES;
+  }
+

```

Comment:
TXT_REGION_COUNT is #defined as 3. Is there any benefit to using TxtIndex as opposed to hardcoding [0], [1], and [2]?

Reply:
For me this is natural, easy to understand, and helps ensure the array index stays with the allocated bounds:

for (TxtIndex = 0; TxtIndex < TXT_REGION_COUNT; TxtIndex++) {
  switch (TxtIndex) {
    case 0:
      // Do zero logic
      break;
    case 1:
      // Do one logic
      break;
    ...
  }
}

Reply:
Updated.